### ISSUE TITLE
Feature Request - Rebuild Image Cache

### ISSUE THREAD

**Issue by Noodleyman** [cubecart/v6#1383]
It's nice that we have a clear cache option, but it would also be nice to have a rebuild cache option. This is especially helpful when dealing with larger stores. For example, let's assume we clear all of the image cache (perhaps due to a skin image size config change).

If you then navigate to a product to edit it, the store will hang whilst it rebuilds the image cache. Eventually it will time out and you get a white screen. You have to repeat the process multiple times until the cache is eventually fully built. 

Alternately, it can also drag down the front end performance if the cache is not fully built yet.

It would be really handy to have the ability to rebuild the entire cache on demand, rather than wait for it to repopulate by user trigger events.

**Comment by havenswift-hosting** [comment]
Care would have to be taken if / when implementing this, that it is done in a controlled way by doing the rebuild in batches and leaving a period of time in between each batch.  This would ensure that server process time & memory limits are not reached but also that the server is not strained.  Clearing image cache, especially on larger stores, should ideally be done in quiet periods

**Comment by Noodleyman** [comment]
In an ideal world, there would be a scheduled job to do this in the background, so the UI doesn't have to wait on threads to finish/update and it will prevent the hanging/performance issues under these scenarios.

**Comment by abrookbanks** [comment]
I just don't think this is feasible. The cache is supposed to gradually build up on demand based on unknown possibilities. I don't know how this would be possible aside from the image cache. The problem is building the image cache all at once would be beyond (in most cases) PHP's resource allocation.

**Comment by Noodleyman** [comment]
To clarify, this ticket is intended to be about image cache only.  I don't think I made that very clear, sorry. 
If a cron job was introduced (optional for users to setup in their hosting control panel), it would be nice to have a setting to toggle image caching between "on demand" and "cron". when set to cron it would disable the CC image caching and depend on the back end cron job to build them.

the cron can run every minute and batch process to remain within resource constraints and configuration limits.

**Comment by abrookbanks** [comment]
Oh ok.. reopening to look again..

**Comment by abrookbanks** [comment]
A cron every minute to check all image caches have been made sounds quite a heavy task.

**Comment by Noodleyman** [comment]
Was simply an example, not the recommended solution. ;)

**Comment by havenswift-hosting** [comment]
Many hosting companies dont allow cron to run every minute anyway but that wouldnt be something CubeCart controls - my suggestion would be a simple script that rebuilds cache for X number of images and exits (X to be determined and happy to help test in real world situation). Once the cron job is setup it would of course continue to run regardless of any setting in CubeCart so I would also suggest the first thing the script does is check the setting is not "rebuild" and also whether all all image caches have been built, then it should email the store owner to tell them to remove the cron task and exit

**Comment by abrookbanks** [comment]
There is no reason why CubeCart can't do this without falling into memory trap problems if the code is written well.

**Comment by havenswift-hosting** [comment]
It would be MUCH better if it was all kept within CubeCart but the rebuild process needs to be slow enough to not affect server performance especially in real world situations with tens or hundreds of stores on a server. Maybe if selecting clear image cache - a message should be displayed notifying store owners that images are not immediately generated !

**Comment by bhsmither** [comment]
Aren't image derivatives rendered on demand? On a per-product just-in-time basis?

**Comment by Noodleyman** [comment]
They are, but take for example category view where multiple products are available. If you are loading that page for the first time on a busy server, it increases the page load time for the initial load. On a large store with thousands of categories, this can lead to performance issues whilst those initial images are being built.

my work around on this kind of stores is to keep forcing the filemanager / product pages to open after image clearing as it rebuilds the cache during those loads. but, it can take many reloads before it's built in full. 

It would be nice to have a process which ensures that the image cache is pre-built, to avoid that delay in the initial page load, or the white screen/php timeout error in the admin area when cache is not yet fully built.
